src/main/java/com/elytradev/concrete/inventory/gui/widget/WPanel.java

```diff
@@ -37,24 +37,180 @@
+	 * @see #getWidget
+	 */
+	public int getWidgetCount() {
+		return children.size();
+	}
+	
+	/**
+	 * Gets the nth widget in this panel.
+	 *
+	 * @param n the index of the widget to get
+	 * @return the n<sup>th</sup> widget in this panel
+	 */
+	public WWidget getWidget(int n) {
+		return children.get(n);
+	}
+	
+	/**
+	 * Appends the specified widget to the end of this panel.
+	 * <p>
+	 * This method changes layout-related information, and therefore,
+	 * invalidates the panel hierarchy. If the panel has already been
+	 * validated, the hierarchy must be re-validated thereafter in order to
+	 * properly display the added widget.
+	 *
+	 * @param w the widget to be added
+	 * @see #invalidate
+	 * @see #validate
+	 */
+	public void add(WWidget w) {
+		add(w, -1);
+	}
+	
+	/**
+	 * Appends the specified widget to this panel at the given position.
+	 * <p>
+	 * This method changes layout-related information, and therefore,
+	 * invalidates the panel hierarchy. If the panel has already been
+	 * validated, the hierarchy must be re-validated thereafter in order to
+	 * properly display the added widget.
+	 *
+	 * @param w     the widget to be added
+	 * @param index the position at which to insert the widget, or
+	 *              <code>-1</code> to append the widget to the end
+	 * @see #remove
+	 * @see #invalidate
+	 * @see #validate
+	 */
+	public void add(WWidget w, int index) {
+		if (index != -1) {
+			children.add(index, w);
+		} else {
+			children.add(w);
 		}
-		valid = true;
+		w.setParent(this);
+		invalidate();
+	}
+	
+	/**
+	 * Removes the widget, specified by <code>index</code>, from this panel.
+	 * <p>
+	 * This method changes layout-related information, and therefore,
+	 * invalidates the panel hierarchy. If the panel has already been
+	 * validated, the hierarchy must be re-validated thereafter in order to
+	 * reflect the changes.
+	 *
+	 * @param index the index of the widget to be removed
+	 * @see #add
+	 * @see #invalidate
+	 * @see #validate
+	 * @see #getWidgetCount
+	 */
+	public void remove(int index) {
+		children.remove(index);
+		invalidate();
 	}
 	
+	/**
+	 * Removes the specified widget from this panel.
+	 * <p>
+	 * This method changes layout-related information, and therefore,
+	 * invalidates the panel hierarchy. If the panel has already been
+	 * validated, the hierarchy must be re-validated thereafter in order to
+	 * reflect the changes.
+	 *
+	 * @param w the widget to be removed
+	 * @see #add
+	 * @see #invalidate
+	 * @see #validate
+	 */
 	public void remove(WWidget w) {
 		children.remove(w);
-		valid = false;
+		invalidate();
+	}
+	
+	/**
+	 * Causes this panel to be sized to fit the preferred size of its widgets.
+	 * The resulting width and height of the panel are automatically enlarged
+	 * if either of dimensions is less than the minimum size as specified by
+	 * the previous call to the {@code setMinimumSize} method.
+	 *
+	 * @see #setMinimumSize
+	 */
+	public void pack() {
```

Comment:
Yes, yes I did. I'll clean up the documentation since you think it's too verbose.